src/mlpack/tests/main_tests/gmm_train_test.cpp

```diff
@@ -0,0 +1,377 @@
+                      gmm1->Component(sortedIndices[k]).Mean()) > 1e-50 ||
+                  arma::norm(gmm->Component(sortedIndices[k]).Covariance() -
+                      gmm1->Component(sortedIndices[k]).Covariance()) > 1e-50);
+}
+
+// Ensure that tolerance affects the final result.
+BOOST_AUTO_TEST_CASE(GmmTrainToleranceTest)
+{
+  arma::mat inputData;
+  if (!data::Load("vc2.csv", inputData))
+    BOOST_FAIL("Unable to load train dataset vc2.csv!");
+
+  SetInputParam("input", inputData);
+  SetInputParam("gaussians", (int) 2);
+  SetInputParam("tolerance", (double) 1e-8);
+
+  mlpack::math::FixedRandomSeed();
+  mlpackMain();
+
+  GMM* gmm = CLI::GetParam<GMM*>("output_model");
+
+  ResetGmmTrainSetting();
+
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("gaussians", (int) 2);
+  SetInputParam("tolerance", (double) 10);
+
+  mlpack::math::FixedRandomSeed();
+  mlpackMain();
+
+  GMM* gmm1 = CLI::GetParam<GMM*>("output_model");
+
+  arma::uvec sortedIndices = sort_index(gmm->Weights());
+
+  for (size_t k = 0; k < sortedIndices.n_elem; k++)
+    BOOST_REQUIRE(arma::norm(gmm->Component(sortedIndices[k]).Mean() -
+                      gmm1->Component(sortedIndices[k]).Mean()) > 1e-50 ||
+                  arma::norm(gmm->Component(sortedIndices[k]).Covariance() -
+                      gmm1->Component(sortedIndices[k]).Covariance()) > 1e-50);
+}
+
+// Ensure that saved model can be used again.
+BOOST_AUTO_TEST_CASE(GmmTrainModelReuseTest)
+{
+  arma::mat inputData(5, 10, arma::fill::randu);
+
+  SetInputParam("input", inputData);
+  SetInputParam("gaussians", (int) 2);
+
+  mlpackMain();
+
+  GMM* gmm = CLI::GetParam<GMM*>("output_model");
+
+  SetInputParam("input_model", gmm);
+
+  CLI::GetSingleton().Parameters()["input"].wasPassed = false;
+
+  SetInputParam("input", inputData);
+
+  mlpackMain();
+
+  GMM* gmm1 = CLI::GetParam<GMM*>("output_model");
+
+  SetInputParam("input_model", gmm1);
+
+  CLI::GetSingleton().Parameters()["input"].wasPassed = false;
+
+  SetInputParam("input", std::move(inputData));
+
+  mlpackMain();
+
+  GMM* gmm2 = CLI::GetParam<GMM*>("output_model");
+
+  BOOST_REQUIRE_EQUAL(gmm1, gmm2);
+}
+
+// Ensure that Gmm's covariances are diagonal when diagonal_covariance is true.
+BOOST_AUTO_TEST_CASE(GmmTrainDiagCovariance)
+{
+  arma::mat inputData(5, 10, arma::fill::randu);
+
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("gaussians", (int) 2);
+  SetInputParam("diagonal_covariance", true);
+
+  mlpackMain();
+
+  GMM* gmm = CLI::GetParam<GMM*>("output_model");
+
+  arma::uvec sortedIndices = sort_index(gmm->Weights());
+
+  for (size_t k = 0; k < sortedIndices.n_elem; k++)
+  {
+    arma::mat diagCov(gmm->Component(sortedIndices[k]).Covariance());
+      for (size_t i = 0; i < diagCov.n_rows; i++)
+        for (size_t j = 0; j < diagCov.n_cols; j++)
+          if (i != j && diagCov(i, j) != (double) 0)
+            BOOST_FAIL("Covariance Are Not Diagonal");
+  }
+}
+
```

Comment:
I can't think of a way to test when `no_force_positive` is given, but I guess the best we can do is just make sure that there is no crash when `no_force_positive` is specified.  (We can't test whether it changes the result, because most of the time, it won't significantly change the result.)

Reply:
I have added the test you can take a look. :)

Reply:
Looks good, thanks!